control-plane/tenancy/namespace/namespace_controller.go

```diff
@@ -0,0 +1,138 @@
+	// namespaces created while mirroring. For example, if it is set to "k8s-",
+	// then the k8s `default` namespace will be mirrored in Consul's
+	// `k8s-default` namespace.
+	MirroringK8sPrefix string
+
+	// CrossNamespaceACLPolicy is the name of the ACL policy to attach to
+	// any created Consul namespaces to allow cross namespace service discovery.
+	// Only necessary if ACLs are enabled.
+	CrossNamespaceACLPolicy string
+
+	Log logr.Logger
+}
+
+// Reconcile reads a Kubernetes Namespace and reconciles the mapped namespace in Consul.
+func (r *Controller) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var namespace corev1.Namespace
+
+	// Ignore the request if the namespace should not be synced to consul.
+	if common.ShouldIgnore(req.Name, r.DenyK8sNamespacesSet, r.AllowK8sNamespacesSet) {
+		return ctrl.Result{}, nil
+	}
+
+	// Create a gRPC resource service client
+	resourceClient, err := consul.NewResourceServiceClient(r.ConsulServerConnMgr)
+	if err != nil {
+		r.Log.Error(err, "failed to create Consul resource service client", "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// Re-read the k8s namespace object
+	err = r.Client.Get(ctx, req.NamespacedName, &namespace)
+
+	consulAP := r.ConsulClientConfig.APIClientConfig.Partition
+	consulNS := r.consulNamespaceFor(req.Name)
+
+	// If the namespace object has been deleted (we get an IsNotFound error),
+	// we need to remove the Namespace from Consul.
+	if k8serrors.IsNotFound(err) {
+		// if we are using a destination namespace, NEVER delete it.
+		if !r.MirroringK8s {
+			return ctrl.Result{}, nil
+		}
+
+		if err := EnsureDeleted(ctx, resourceClient, consulAP, consulNS); err != nil {
+			return ctrl.Result{}, fmt.Errorf("error deleting consul namespace: %w", err)
+		}
+
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		r.Log.Error(err, "failed to get k8s namespace", "name", req.Name)
+		return ctrl.Result{}, err
+	}
+
+	// k8s namespace found, so make sure it is mapped correctly and exists in Consul.
+	r.Log.Info("retrieved", "k8s namespace", namespace.GetName())
+
+	// TODO: revisit setting crossNSACLPolicy when V2 ACLS exist
+	if _, err := EnsureExists(ctx, resourceClient, consulAP, consulNS); err != nil {
+		r.Log.Error(err, "error checking or creating consul namespace", "namespace", consulNS)
+		return ctrl.Result{}, fmt.Errorf("error checking or creating consul namespace: %w", err)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager registers this controller with the manager.
+func (r *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&corev1.Namespace{}).
+		Complete(r)
+}
+
+// consulNamespaceFor returns the Consul destination namespace for a provided Kubernetes namespace
+// depending on Consul Namespaces being enabled and the value of namespace mirroring.
+func (r *Controller) consulNamespaceFor(kubeNamespace string) string {
```

Comment:
didnt we want just 1:1 mapping for Consul and Kubernetes namespace in v2?

Reply:
as in Kubernetes Namespace name == Consul Namespace name?

Reply:
Removed since mapping is now straight 1:1